Add Service Tagging Capability

XedinUnknown opened the pull request:
Now you can tag your services, like so:

```php
[
    'serviceA' =>
        /** @tag letters */
        fn (): string => 'A',
    'serviceB' =>
        /**
         * @tag letters
         */
        function (): string {
            return 'B';
        },
    'serviceC' => function (ContainerInterface $c): string {
        var_dump($c->get('letters'));
    },
];
```

For every unique tag, an extension is added that resolves tagged services and adds them to the list. The list is ensured by also creating an empty list factory with the same name.

XedinUnknown committed:
Add `TaggingServiceProvider`

XedinUnknown committed:
Adjust docblock to test single-line

mecha commented on src/TaggingServiceProvider.php:
```diff
@@ -0,0 +1,121 @@
+     */
+    protected function getTagsFromDocBlock(string $docBlock): array
+    {
+        $regex = '#(@tag\s*(?P<tags>[^\s]+))#';
```
This regex will catch `@tags` that aren't in the "beginning" of the line, such as:

```php
/** There is a misplaced @tag in this comment. */

/**
 * There is a misplaced @tag in this comment.
 */
```

I added to services to the test to confirm this, `serviceY` and `serviceZ` for the above two scenarios, and both were detected.

Might I suggest this regex instead:

```suggestion
        $regex = '#^\s*/?\**\s*(@tag\s*(?P<tags>[^\s]+))#m';
```

I could also commit and push my local changes, if you want. Or submit a PR targeting this branch.

XedinUnknown replied:
If it's possible - please feel free to add the rest of your changes e.g. the test to this very branch directly.

XedinUnknown committed:
Ensure tag appears first on docblock line

Co-authored-by: Miguel Muscat <miguelmuscat93@gmail.com>

mecha committed:
Add test for misplaced tags

XedinUnknown commented on src/TaggingServiceProvider.php:
```diff
@@ -113,7 +113,7 @@ protected function indexTags(): void
     protected function getTagsFromDocBlock(string $docBlock): array
     {
-        $regex = '#(@tag\s*(?P<tags>[^\s]+))#';
+        $regex = '#^\s*/?\**\s*(@tag\s*(?P<tags>[^\s]+))#m';
```
I guess, the `m` modifier means that if the tag spans multiple lines, this would still work, right? Although, I don't see how that could be: a tag name is a service name, and should not contain spaces.

mecha replied:
Actually, the `m` modifier simply changes the `^` and `$` tokens to match the start/end of the line, rather than the start/end of the string ([source](https://www.php.net/manual/en/reference.pcre.pattern.modifiers.php)). The modifier doesn't really change anything. If they can span multiple lines with the modifier, they can span multiple lines without it.

As far as I can tell, new lines are only possible where the `\s` whitespace tokens are, and the only one that I see that would affect tag detection is the token between the `@tag` sequence and the service name capture group. Though in most sensible cases, if the `@tag` string is on a separate line from the service name, there's probably going to be asterisks in between them anyway, so it would fail to match with the regex pattern.

```php
/**
 * @tag
 * my_service
 */
```

You'd have to go out of your way to accommodate it, which now that I look at it makes it seem almost reasonable that it would match with the regex:

```php
/**
 * @tag
    my_service
 */
```

I personally wouldn't lose sleep over this 😅

XedinUnknown replied:
Thank you kindly, sir, for the explanation! 🙏 That is exactly the kind of case I had in mind. And yes, if I really do go out of my way so far in order to do this, then I probably meant to do exactly this for whatever reason - maybe to satisfy a `phpcs` line length limitation with minimal casualties.

So, ya, I pronounce this good and done.

XedinUnknown commented:
Looking good!

@biont has also had a look at this, and I have received positive feedback. 

Thanks, all!

XedinUnknown committed:
Update PHPStorm configs

XedinUnknown committed:
Update deps

XedinUnknown committed:
Fix a potential bug reported by Psalm

The default calculation should only have effect when `$length` has the default value of `null`, not when it's just falsy.

XedinUnknown committed:
Remove redundant call

XedinUnknown committed:
Add a Psalm baseline file

This suppresses a bunch of errors en-masse. They are very difficult to satisfy.

See also https://github.com/vimeo/psalm/issues/10922

XedinUnknown committed:
Migrated PHPUnit configuration to newer format

XedinUnknown committed:
Reformat config file

XedinUnknown committed:
Upgrade dependency installation step technique

XedinUnknown committed:
Cache Composer dependencies

XedinUnknown committed:
Build also on PHP 8.3

XedinUnknown committed:
Update changelog

XedinUnknown committed:
Update readme

Added description of Service Tagging feature.

XedinUnknown committed:
Upgrade GH Action. This increases Composer to v2

XedinUnknown committed:
Trying to reset cache 1/2

XedinUnknown committed:
Remove config preventing Composer v2
